actioncable/lib/action_cable/channel/test_case.rb

```diff
@@ -0,0 +1,207 @@
+require "active_support/test_case"
+require "active_support/core_ext/hash/indifferent_access"
+require "json"
+
+module ActionCable
+  module Channel
+    class NonInferrablChannelError < ::StandardError
+      def initialize(name)
+        super "Unable to determine the channel to test from #{name}. " +
+          "You'll need to specify it using tests YourChannel in your " +
+          "test case definition"
+      end
+    end
+
+    # Stub `sream_from` to track streams for the channel
+    module TestStreams
+      def stream_from(broadcasting, callback = nil, coder: nil)
+        streams << broadcasting
+      end
+
+      def stop_all_streams
+        @_streams = []
+      end
+
+      def streams
+        @_streams ||= []
+      end
+    end
+
+    class TestSubscriptions < ActionCable::Connection::Subscriptions
+      delegate :count, to: :identifiers
+
+      def add(channel, params)
+        identifier = {
+          channel: channel.name
+        }.merge(params).to_json
+
+        subscription = channel.new(connection, identifier, params.with_indifferent_access)
+        subscription.singleton_class.include(TestStreams)
+        subscriptions[identifier] = subscription
+        subscription.subscribe_to_channel
+        subscription
+      end
+    end
+
+    class TestConnection
+      attr_reader :transmissions, :identifiers, :subscriptions
+
+      def initialize(identifiers = {})
+        @transmissions = []
+
+        identifiers.each do |identifier, val|
+          define_singleton_method(identifier) { val }
+        end
+
+        @identifiers = identifiers.keys
+        @subscriptions = TestSubscriptions.new(self)
+      end
+
+      def transmit(cable_message)
+        transmissions << cable_message.with_indifferent_access
+      end
+
+      def logger
+        ::ActionCable.server.config.logger
+      end
+    end
+
+    # Superclass for ActionCable channels functional tests. Functional tests allow you to
```

Comment:
We've also moved away from the functional term in controller tests. So let's just say channel tests.